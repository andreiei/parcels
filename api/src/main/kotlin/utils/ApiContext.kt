package utils

import services.ParcelRepository
import services.RoutesService

data class ApiContext(
    val parcelRepository: ParcelRepository = ParcelRepository(),
    val routesService: RoutesService = RoutesService(),
)
