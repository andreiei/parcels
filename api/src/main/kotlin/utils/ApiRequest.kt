package utils

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import decodeFromString
import failure.ApiException
import kotlinx.serialization.Serializable

@Serializable
data class ApiRequest(
    val body: String?,
    val path: String,
    val httpMethod: String,
) {
    inline fun <reified T> getBody(): T {
        val rawBody: String = body ?: throw ApiException.BadRequestException("Request body is missing")
        return runCatching {
            decodeFromString<T>(rawBody)
        }
            .getOrElse {
                throw ApiException.BadRequestException("Failed to parse request body into ${T::class.simpleName}")
            }
    }

    fun getCustomPath(): String = "$httpMethod $path"

    companion object {
        fun APIGatewayProxyRequestEvent.toApiRequest(): ApiRequest = ApiRequest(
            body = body,
            path = path,
            httpMethod = httpMethod,
        )
    }
}
