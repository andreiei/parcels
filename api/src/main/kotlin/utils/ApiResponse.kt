package utils

import encodeToString
import kotlinx.serialization.Serializable

@Serializable
data class ApiResponse(
    val body: String?,
    val code: Int,
) {
    companion object {
        inline fun <reified T> withContent(body: T, code: Int = 200): ApiResponse =
            ApiResponse(body = encodeToString(body), code = code)

        fun noContent(): ApiResponse = ApiResponse(body = null, code = 204)
    }
}
