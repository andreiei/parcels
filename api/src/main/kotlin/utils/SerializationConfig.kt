import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer

inline fun <reified T> encodeToString(data: T): String = SerializerConfig.json.encodeToString(serializer(), data)
inline fun <reified T> decodeFromString(data: String): T = SerializerConfig.json.decodeFromString(serializer(), data)

open class SerializerConfig {
    companion object {
        val json: Json =
            Json {
                isLenient = true
                prettyPrint = true
                encodeDefaults = true
                ignoreUnknownKeys = true
            }
    }
}
