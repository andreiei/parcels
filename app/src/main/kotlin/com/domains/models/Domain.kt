package com.domains.models

@JvmInline
value class Domain private constructor(val value: String) {
    override fun toString(): String = value

    companion object {
        private val domainRegex = Regex("""^(?!-)(?!.*--)(?!.*\.\.)([a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}$""")
        fun from(raw: String): Domain? {
            val cleaned: String = raw.trim().lowercase()
            return if (domainRegex.matches(cleaned)) Domain(cleaned) else null
        }
    }
}
