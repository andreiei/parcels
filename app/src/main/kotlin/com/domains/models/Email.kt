package com.domains.models

@JvmInline
value class Email private constructor(val value: String) {
    override fun toString(): String = value

    fun isValid(): Boolean {
        val emailRegex =
            Regex("""^[a-zA-Z0-9._%+-]+@(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}$""")
        return emailRegex.matches(value)
    }

    fun getDomain(): Domain? = Domain.from(value.substringAfter('@'))

    companion object {
        fun from(value: String): Email? {
            val email = Email(value.trim())
            return if (email.isValid()) email else null
        }
    }
}
