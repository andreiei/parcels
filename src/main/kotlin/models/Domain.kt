package models

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class Domain private constructor(val value: String) {
    override fun toString(): String = value

    companion object {
        private val domainRegex = Regex("""^(?!-)(?!.*--)(?!.*\.\.)([a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}$""")
        fun from(raw: String): Domain? {
            val cleaned: String = raw.trim().lowercase()
            return if (domainRegex.matches(cleaned)) Domain(cleaned) else null
        }

        fun getDomain(email: Email): Domain? = email.value.substringAfter('@').let { from(it) }
    }
}
