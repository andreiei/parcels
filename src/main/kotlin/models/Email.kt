package models

import org.apache.commons.validator.routines.EmailValidator

@JvmInline
value class Email private constructor(val value: String) {
    override fun toString(): String = value

    companion object {
        fun from(value: String): Email? {
            val email = Email(value.trim())
            return if (EmailValidator.getInstance().isValid(value)) email else null
        }
    }
}
