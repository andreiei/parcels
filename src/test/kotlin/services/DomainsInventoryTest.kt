package services

import io.kotest.core.spec.style.ShouldSpec
import models.Domain
import utils.shouldMatchSnapshot

class DomainsInventoryTest : ShouldSpec() {
    init {
        should("get N amount of domains") {
            val domainsInventory = DomainsInventory()

            listOf(
                createDomains("gmail.com", 5),
                createDomains("yahoo.com", 12),
                createDomains("hotmail.com", 8),
            )
                .flatten()
                .forEach {
                    domainsInventory.add(it)
                }

            domainsInventory
                .get()
                .map { "${it.domain}: ${it.count}" }
                .shouldMatchSnapshot(this)
        }
    }

    private fun createDomains(name: String, amount: Int): List<Domain> =
        (1..amount).mapNotNull { Domain.from(name) }
}
