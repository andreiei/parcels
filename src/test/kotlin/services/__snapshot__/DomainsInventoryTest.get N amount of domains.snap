[
    "yahoo.com: 12",
    "hotmail.com: 8",
    "gmail.com: 5"
]